### PR TITLE
Spark: Perf improvement for query/changefeed - Pooling RowSerializers

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
Open Items: Still working on a couple of new unit tests for the RowSerializerPool class

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  }
+
+  def returnSerializerToPool(schema: StructType, serializer: ExpressionEncoder.Serializer[Row]): Boolean = {
+    schemaScopedSerializerMap.get(schema) match {
+      case Some(objectPool) => objectPool.returnSerializer(serializer)
+      case None =>
+        val newQueue = new RowSerializerQueue()
+        newQueue.returnSerializer(serializer)
+        !schemaScopedSerializerMap.putIfAbsent(schema, newQueue).isDefined
```
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:50]: Should the put be happening on the borrow instead?

**Review by ealsur — APPROVED** [review]

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ */
+private object RowSerializerPool {
+  val MaxPooledSerializerCount = 256
+  private[this] val cleanUpIntervalInSeconds = 300
+  private[this] val expirationIntervalInSeconds = 1800
+  private[this] val schemaScopedSerializerMap =
+    new TrieMap[StructType, RowSerializerQueue]
+  private[this] val executorService = Executors.newSingleThreadScheduledExecutor(SparkUtils.daemonThreadFactory())
+
+  executorService.scheduleWithFixedDelay(
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:31]: do we need to schedule more cleanup if no new row encoder is left?

can we schedule the next cleanup() in getOrCreateSerializer()? basically we can check if there is any cleanup scheduled if not schedule one.

That way if row encoders are not used for a while we don't have schedule new ones.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    () => this.onCleanUp(),
+    cleanUpIntervalInSeconds,
+    cleanUpIntervalInSeconds,
+    TimeUnit.SECONDS)
+
+  def getOrCreateSerializer(schema: StructType): ExpressionEncoder.Serializer[Row] = {
+    schemaScopedSerializerMap.get(schema) match {
+      case Some(objectPool) => objectPool.borrowSerializer(schema)
+      case None => RowEncoder(schema).createSerializer()
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:40]: Don't think this is necessary here - worst case we create one more serializer than necessary - that race condition is acceptable

**Review by moderakh — APPROVED** [review]
lgtm. minor non blocking comment.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:31]: The cleanup only happens every 5 minutes - don't think it is necessary to optimize perf/overhead for it. IMO keeping it as simple as possible is sufficient here.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala]: I don't think functionally it makes a difference - motivation for putting it here was that I think the risk of race conditions (where putIfAbsence would find a concurrently added serializer) is lower when doing it on task completion than at the beginning of a job. That race condition shouldn't cause any issues - so I don't have a strong preference thought. Why owuld you prefer it on borrow?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:31]: we can do later if needed. LGTM.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 1 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala]: Well, it's the pattern I've seen most used, if the creation of the resource is what is expensive, then gate the creation of the resource and avoid extra creations. That way, in a concurrent scenario of multiple partition readers start, you avoid the cost of creation.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:40]: But wasn't that the scenario we were trying to solve? Before, we were creating one serializer per Table or Partition reader, if we don't guard the creation, aren't we falling on the same boat potentially?

[the remark this answers is not in the document]
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala:40]: Discussed offline - The issue is that the Serializer is not thread-safe

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala (hunk 1 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala]: Resolved offline - The issue is the serializer is not thread-safe.

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala]: Discussed offline with Matias - we agree leaving it here is fine. Main misunderstand was that the Serializer isn't thread-safe and so we cannot allow it to be used across PartitionReaders. And also the cost of creating a Serializer was definitely too high when doing it for every Row - but having a few race conditions where we might create a couple more Serializer instances than strictly required isn't the end of the world. The object pooling will help over time still significantly.

**Pull request merged**